### PR TITLE
Bug 1924447: Migration to the new Project structure as per latest operator-sdk

### PR DESCRIPTION
**Pull request by priyanka19-98**
Ref docs: 
https://sdk.operatorframework.io/docs/building-operators/golang/migration/
https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.0.0/

Signed off by: Priyanka Jiandani <pjiandan@redhat.com>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,63 +1,62 @@
...
-	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.0.0-20191016112429-9587704a8ad4
-	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.0.0-20191016114939-2b2b218dc1df
-	k8s.io/kube-proxy => k8s.io/kube-proxy v0.0.0-20191016114407-2e83b6f20229
-	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.0.0-20191016114748-65049c67a58b
-	k8s.io/kubectl => k8s.io/kubectl v0.0.0-20191016120415-2ed914427d51
-	k8s.io/kubelet => k8s.io/kubelet v0.0.0-20191016114556-7841ed97f1b2
-	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.0.0-20191016115753-cf0698c3a16b
-	k8s.io/metrics => k8s.io/metrics v0.0.0-20191016113814-3b1a734dba6e
-	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20191016112829-06bb3c9d77c9
+	k8s.io/api => k8s.io/api v0.20.0
```
Comment (jsafrane) [inline go.mod:34]: I think this should be v0.21.0 (applies to all k8s deps below)

Comment (priyanka19-98) [reply go.mod]: updated

**Comment by gnufied** [comment]
@priyanka19-98 can you link the PR to a BZ? The PR can't merge without a valid BZ.

File: common/names.go [hunk 2]
```diff
@@ -4,12 +4,12 @@ import (
...
 	"os"
 
-	localv1 "github.com/openshift/local-storage-operator/pkg/apis/local/v1"
+	localv1 "github.com/openshift/local-storage-operator/api/v1"
 	corev1 "k8s.io/api/core/v1"
 )
 
 const (
-	defaultDiskMakerImageVersion = "quay.io/openshift/origin-local-storage-diskmaker"
+	defaultDiskMakerImageVersion = "quay.io/pjiandan/local-diskmaker"
```
Comment (gnufied) [inline common/names.go:12]: we should fix these before having final PR?

Comment (priyanka19-98) [reply common/names.go]: fixed

File: common/controller_runtime_utils.go [hunk 3]
```diff
@@ -56,3 +57,12 @@ func GetNodeNameEnvVar() string {
 func GetWatchNameSpaceEnfVar() string {
 	return os.Getenv("WATCH_NAMESPACE")
 }
+
+// GetWatchNamespace returns the namespace the operator should be watching for changes
+func GetWatchNamespace() (string, error) {
```
Comment (jsafrane) [inline common/controller_runtime_utils.go:62]: Would it be possible to merge it with `GetWatchNameSpaceEnfVar` just above? I.e. rework `GetWatchNameSpaceEnfVar` to call `GetWatchNamespace()` (and fix the typo in the function name :-)) or even remove it?

File: controllers/localvolume/api_updater.go [hunk 4]
```diff
@@ -121,27 +121,27 @@ func (s *sdkAPIUpdater) applyClusterRoleBinding(roleBinding *rbacv1.ClusterRoleB
...
+func (s *sdkAPIUpdater) applyStorageClass(ctx context.Context, sc *storagev1.StorageClass) (*storagev1.StorageClass, bool, error) {
+	return applyStorageClass(ctx, s.clientset.StorageV1(), sc)
 }
 
 func (s *sdkAPIUpdater) applyDaemonSet(ds *appsv1.DaemonSet, expectedGeneration int64, forceRollout bool) (*appsv1.DaemonSet, bool, error) {
 	if forceRollout {
 		klog.Infof("Rolling out DaemonSet: %s/%s", ds.Name, ds.Namespace)
 	}
-	return resourceapply.ApplyDaemonSet(s.clientset.AppsV1(), events.NewInMemoryRecorder(componentName), ds, expectedGeneration, forceRollout)
+	return resourceapply.ApplyDaemonSetWithForce(s.clientset.AppsV1(), events.NewInMemoryRecorder(componentName), ds, expectedGeneration, forceRollout)
```
Comment (jsafrane) [inline controllers/localvolume/api_updater.go:132]: Nit: I see half of the `sdkAPIUpdater` methods are not used anywhere, such as `applyDaemonSet`, `getDaemonSet`, `applyClusterRoleBinding`, `applyClusterRole`, `applyRoleBinding`, `applyRole` and maybe more. Can you please remove them in a subsequent PR?

File: controllers/localvolumediscoveryresult_controller.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+// the LocalVolumeDiscoveryResult object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.7.2/pkg/reconcile
+func (r *LocalVolumeDiscoveryResultReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = r.Log.WithValues("localvolumediscoveryresult", req.NamespacedName)
+
+	// your logic here
```
Comment (jsafrane) [inline controllers/localvolumediscoveryresult_controller.go:53]: Is this reconciler necessary? It seems it does not do anything useful

File: controllers/localvolume/localvolume_controller.go [hunk 6]
```diff
@@ -325,3 +373,55 @@ func remove(list []string, s string) []string {
...
+				// skip LocalVolumeSet owned PVs
+				ownerKind, found := pv.Labels[common.PVOwnerKindLabel]
+				if ownerKind != localv1.LocalVolumeKind || !found {
+					return []reconcile.Request{}
+				}
+				req := reconcile.Request{NamespacedName: types.NamespacedName{Name: ownerName, Namespace: ownerNamespace}}
+				return []reconcile.Request{req}
+			})).
+		//	Owns(&appsv1.DaemonSet{}).
+		//Owns(&corev1.PersistentVolume{}).
```
Comment (jsafrane) [inline controllers/localvolume/localvolume_controller.go:425]: Are these comments necessary?

File: controllers/localvolumeset/localvolumeset_controller.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+// Reconcile reads that state of the cluster for a LocalVolumeSet object and makes changes based on the state read
+// and what is in the LocalVolumeSet.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *LocalVolumeSetReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	result, err := r.reconcile(ctx, request)
+	// sets conditions based on the exit status of reconcile
+	return r.addAvailabilityConditions(ctx, request, result, err)
+	//	return ctrl.Result{}, nil
```
Comment (jsafrane) [inline controllers/localvolumeset/localvolumeset_controller.go:72]: Please remove the comment

File: controllers/suite_test.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
+/*
```
Comment (jsafrane) [inline controllers/suite_test.go:1]: Is this file necessary? Or why is it commented out?

File: hack/boilerplate.go.txt [hunk 9]
```diff
@@ -0,0 +1,15 @@
+/*
+Copyright 2021.
```
Comment (jsafrane) [inline hack/boilerplate.go.txt:2]: Can you update the boilerplate with copyright owner (dunno, "Copyright 2021 The Local Storage Operator authors"?) and regenerate everything?

cc @gnufied @rohantmp @sp98 what do you think would be appropriate here?

**Review by jsafrane — COMMENTED** [review]
Went through the first commit.

**Comment by jsafrane** [comment]
Trying to fix the unit tests in https://github.com/openshift/release/pull/18879.

**Comment by jsafrane** [comment]
Sorry, did not notice your CI PR.

File: api/v1/localvolume_types.go [hunk 10]
```diff
@@ -70,11 +86,9 @@ type LocalVolumeStatus struct {
 	Generations []operatorv1.GenerationStatus `json:"generations,omitempty"`
 }
 
-// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-
-// LocalVolume is the Schema for the localvolumes API
-// +kubebuilder:subresource:status
```
[the remark this answers is not in the document]
Comment (priyanka19-98) [reply api/v1/localvolume_types.go]: added

File: common/controller_runtime_utils.go (hunk 3 above)
Comment (priyanka19-98) [reply common/controller_runtime_utils.go:57]: removed

File: controllers/localvolume/api_updater.go (hunk 4 above)
Comment (priyanka19-98) [reply controllers/localvolume/api_updater.go]: removed

File: controllers/localvolumediscoveryresult_controller.go (hunk 5 above)
Comment (priyanka19-98) [reply controllers/localvolumediscoveryresult_controller.go]: removed

File: controllers/localvolume/localvolume_controller.go (hunk 6 above)
Comment (priyanka19-98) [reply controllers/localvolume/localvolume_controller.go]: removed

File: controllers/localvolumeset/localvolumeset_controller.go (hunk 7 above)
Comment (priyanka19-98) [reply controllers/localvolumeset/localvolumeset_controller.go]: removed

File: controllers/suite_test.go (hunk 8 above)
Comment (priyanka19-98) [reply controllers/suite_test.go:1]: This file was generated by `operator-sdk` . I kept it because I thought we might need to move e2e-tests in this file in the future. Currently we are using `operator-framework` . Let me know if it needs to be removed.

File: hack/boilerplate.go.txt (hunk 9 above)
Comment (sp98) [reply hack/boilerplate.go.txt:2]: `Copyright 2021 The Local Storage Operator Authors` looks good to me.

**Comment by sp98** [comment]
Now we have crds in both `config/crd/bases` and `deploy` directory.  Is that expected?  Looks like redundancy.  CRDs should be moved to `config/crd/bases` according to the operator-sdk migration doc.

**Comment by priyanka19-98** [comment]
> @priyanka19-98 Now we have crds in both `config/crd/bases` and `deploy` directory. Is that expected? Looks like redundancy. CRDs should be moved to `config/crd/bases` according to the operator-sdk migration doc.
> 
> Also there is redundant manifests directory. `config/manifests/` and the root `manifests/`

Thanks for pointing this. I ll update The PR.

File: hack/boilerplate.go.txt (hunk 9 above)
Comment (priyanka19-98) [reply hack/boilerplate.go.txt:2]: @jsafrane  @sp98  do you know how could I regenerate everything after editing `boilerplate.go.txt`?

**Comment by gnufied** [comment]
Moving the `manifests` directory inside `config/` directory IMO will be problematic. IIRC ART tooling depends on finding that directory with provided structure and if we simply move it away, it will stop building our images.

@priyanka19-98 have we talked to ART team and if they support new directory structure?

**Comment by gnufied** [comment]
I tried to build a index image from files present in `opm-bundle` directory and got:

```
ERRO[0001] permissive mode disabled                      bundles="[quay.io/gnufied/local-storage-bundle:jul1-17]" error="error checking provided apis in bundle local-storage-operator.v4.8.0: couldn't find local
.storage.openshift.io/v1alpha1/LocalVolumeSet (localvolumesets) in bundle. found: map[local.storage.openshift.io/v1/LocalVolume (localvolumes):{} local.storage.openshift.io/v1alpha1/LocalVolumeDiscovery (localv
olumediscoveries):{} local.storage.openshift.io/v1alpha1/LocalVolumeDiscoveryResult (localvolumediscoveryresults):{}]"                                                                                           
Error: error checking provided apis in bundle local-storage-operator.v4.8.0: couldn't find local.storage.openshift.io/v1alpha1/LocalVolumeSet (localvolumesets) in bundle. found: map[local.storage.openshift.io/v
1/LocalVolume (localvolumes):{} local.storage.openshift.io/v1alpha1/LocalVolumeDiscovery (localvolumediscoveries):{} local.storage.openshift.io/v1alpha1/LocalVolumeDiscoveryResult (localvolumediscoveryresults):
{}]                                                                                                      
Usage:                                                                                                                                                                                                            
  opm index add [flags]                                   
```

File: Makefile [hunk 11]
```diff
@@ -137,58 +160,106 @@ rm -rf $$TMP_DIR ;\
...
+# # This recipe invokes 'opm' in 'semver' bundle add mode. For more information on add modes, see:
+# # https://github.com/operator-framework/community-operators/blob/7f1438c/docs/packaging-operator.md#updating-your-existing-operator
+# .PHONY: catalog-build
+# catalog-build: opm ## Build a catalog image.
+# 	$(OPM) index add --container-tool docker --mode semver --tag $(CATALOG_IMG) --bundles $(BUNDLE_IMGS) $(FROM_INDEX_OPT)
+
+# # Push the catalog image.
+# .PHONY: catalog-push
+# catalog-push: ## Push a catalog image.
+# 	$(MAKE) docker-push IMG=$(CATALOG_IMG)
```
Comment (gnufied) [inline Makefile:217]: Why is all this stuff commented out?

Comment (priyanka19-98) [reply Makefile:217]: this is the stuff generated by operator-sdk . I commented it out to check if we need this stuff in Makefile or we could remove it. There are some commands for creating `bundle` image,` catalog-source` etc. But for bundle stuff we use hack/sync_bundle anyway.

**Comment by priyanka19-98** [comment]
> I tried to build a index image from files present in `opm-bundle` directory and got:
> 
> ```
> ERRO[0001] permissive mode disabled                      bundles="[quay.io/gnufied/local-storage-bundle:jul1-17]" error="error checking provided apis in bundle local-storage-operator.v4.8.0: couldn't find local
> .storage.openshift.io/v1alpha1/LocalVolumeSet (localvolumesets) in bundle. found: map[local.storage.openshift.io/v1/LocalVolume (localvolumes):{} local.storage.openshift.io/v1alpha1/LocalVolumeDiscovery (localv
> olumediscoveries):{} local.storage.openshift.io/v1alpha1/LocalVolumeDiscoveryResult (localvolumediscoveryresults):{}]"                                                                                           
> Error: error checking provided apis in bundle local-storage-operator.v4.8.0: couldn't find local.storage.openshift.io/v1alpha1/LocalVolumeSet (localvolumesets) in bundle. found: map[local.storage.openshift.io/v
> 1/LocalVolume (localvolumes):{} local.storage.openshift.io/v1alpha1/LocalVolumeDiscovery (localvolumediscoveries):{} local.storage.openshift.io/v1alpha1/LocalVolumeDiscoveryResult (localvolumediscoveryresults):
> {}]                                                                                                      
> Usage:                                                                                                                                                                                                            
>   opm index add [flags]                                   
> ```

ahh. thanks for pointing out. fixed this.
